### ISSUE TITLE
XML Conversion

### ISSUE THREAD

**Issue by shitwolfymakes** [kikotheexile/Endless-Sky-Civil-War#6]
**Introduction**
Given that this project is a chance to fix the things that Endless Sky is having trouble fixing, we have the opportunity to demo this. Converting all the data files to XML will make it easier to develop tools to accelerate content creation (like [ESMB](https://github.com/shitwolfymakes/Endless-Sky-Mission-Builder)), make it easier to extend the game itself, and is just better practice than what already exists in the source code.

**Describe the solution you'd like**
1. Rewrite the [Data Format](https://github.com/endless-sky/endless-sky/wiki/DataFormat) in XML

2. Update the source code to read in the new XML files

3. Convert all game data files to the new XML formatting

**Additional context**
1. I will attach examples of what the XML data files will look like as I get to them

2. Per my announcement on the Discord about this, there is a bounty for whoever successfully completes this pull request.

**Comment by MCOfficer** [comment]
i like the idea here, but i would recommend not using xml of all things. i know not everyone is into my yaml-hype, but *please*, let's not limit ourself to a language that is in fact older than myself. i know i'm biased against xml, but i'd still like a discussion about different data formats.

**Comment by kikotheexile** [comment]
> 
> 
> i like the idea here, but i would recommend not using xml of all things. i know not everyone is into my yaml-hype, but _please_, let's not limit ourself that is in fact older than myself. i know i'm biased against xml, but i'd still like a discussion about different data formats.

The conversion to XML is to help with tools like the mission builder, and also to make the game easier for people to find syntax errors in, which is why I support it. 

**H O W E V E R**

Any and all changes that let the engine read from different file types are great by me. I would still like the base game files to largely be in xml or txt, but am down with yaml. Hell, I would even love to see the AI themselves scripted in Lua or *something better suited to handle them than a C language* so that they can be better tweaked for responsiveness as development progresses.

**Comment by kikotheexile** [comment]
Tl;dr "why not both?"

**Comment by lukearndt** [comment]
XML is not a nice language. I would greatly prefer something more modern, like YAML or JSON.

**Comment by kikotheexile** [comment]
I'm down for YAML but would prefer not to use JSON

This conversion is strictly to make tools more viable and accessible for modders. Simply allowing the engine to read from any format we want will allow content creators to blossom without having to get overwhelmed by our text wall system. 

The XML conversion would help, specifically, the ESMB.

**Comment by lukearndt** [comment]
There are good reasons to keep the number of supported languages as low as possible:

- Having additional, unnecessary languages would make it more difficult for people to understand the data files. Instead of needing to know how one language works, now they have to understand 2, 3, 4, or whatever.
- Adding languages also increases the complexity of the parsing system, which makes that code more difficult to maintain.

If we're looking to add a standard language, I think we should pick just one and stick to it. XML is a poor choice in my opinion because it's proven itself to not be very readable / editable by humans.

If JSON is a no-go, then YAML is probably the best option out there at the moment.

**Comment by kikotheexile** [comment]
XML is perfectly readable by humans. I find it easier to read than standard txt files and yaml. JSON is worse than both. 

Tbh we could even just create our own text formatting system if we wanted to. But wolfy is willing to pay like, an entire part time minimum wage paycheck for XML so he has an easier time creating tools.

If all tools are made to read and write XML then that nullifies your concerns, as humans won't usually have to read XML period, the tools will do all that work for them, and everything else can still be handled by text

**Comment by lukearndt** [comment]
> XML is perfectly readable by humans. I find it easier to read than standard txt files and yaml. JSON is worse than both.

It seems we have the exact opposite opinions on which languages are nicer to read than others. I think we'll have to agree to disagree, since readability is subjective anyway.

> If all tools are made to read and write XML then that nullifies your concerns, as humans won't usually have to read XML period, the tools will do all that work for them, and everything else can still be handled by text

- What about the case where human contributors want to read, understand, or edit a data file that was created by a tool?

- What about when it comes time to review a data file in a pull request?

**Comment by kikotheexile** [comment]
Documentation is always important and should never be optional. Should be easy enough to implement

**Comment by kikotheexile** [comment]
To clarify, because i was in a rush with the last reply and really dont feel like doing an edit:

If we explain everything in such a way that a boomer and/or kid who can barely read can understand it with proper notation - which is something I want anyway - then we shouldn't have any problems with new people coming in and trying to read the code. The documentation should quite literally teach them how to read each individual piece of XML just... right there, in a notation. Zero guess work that way for merging/reviewing for pull requests, zero issues for people who want to get in and edit the files without a knowledge of XML. 

A lot of files, such as the map.txt, weapons, outfits, so on should and likely will still be handled in .txt *until* a tool is made for them, too. Meaning that conventional understanding of spacing will work just fine for these people. It's not like we are converting the game to being run in js, c#, and python. The point is literally to make outside tools read and parse this easier for ease of editing and modding. As tools become available, we package them with official releases.

**Comment by shitwolfymakes** [comment]
So my thought process for this is two-fold:

1. Missions would be better as JSON files, but conversations would be a pain to store in JSON format
2. Conversations are best stored in an XML format

Building on this, it's clear that storing the non-conversation data of a mission is just as easy in XML as in JSON, but not the other way around. Also, the whole point of converting to XML from the current data format is so we can declare EXACTLY when conversations, missions, Triggers, Conditions, etc. end. YAML doesn't have closing brackets, which means we're forced to count whitespace (which I do already and was a massive pain in the ass to start working). 

Furthermore, there is vastly more and higher quality libraries for XML in both C++ and Python(XML support is part of the Python standard library) than for YAML.

**Comment by MCOfficer** [comment]
> 
> 
> So my thought process for this is two-fold:
> 
>   1. Missions would be better as JSON files, but conversations would be a pain to store in JSON format
>   2. Conversations are best stored in an XML format
> 
> 
> Building on this, it's clear that storing the non-conversation data of a mission is just as easy in XML as in JSON, but not the other way around. Also, the whole point of converting to XML from the current data format is so we can declare EXACTLY when conversations, missions, Triggers, Conditions, etc. end. YAML doesn't have closing brackets, which means we're forced to count whitespace (which I do already and was a massive pain in the ass to start working).
> 
> Furthermore, there is vastly more and higher quality libraries for XML in both C++ and Python(XML support is part of the Python standard library) than for YAML.

YAML 1.2 is a superset of JSON - anything you can express in JSON, you can express in YAML as well (and libraries can parse it). Furthermore, it has such neat things as multiline-values, one of the big kickers for conversations.

For libraries AFAIK there's libyaml for c++ and ruamel-yaml for python, both of which support YAML 1.2.

As for whitespaces, i'm not all that concerned, a proper editor can provide all the help you need for that - see python. but that's personal preference to some degree. i'm not sure what other than human error margin that concern might be about.

**Comment by kikotheexile** [comment]
Let me put it this way:

- I don't care what language we use as long as it isn't the default of JSON, because for some reason JSON gives me headaches 
- YAML may be a superset of JSON but it's actually easier for me to read without giving me headaches. 
- I am already well versed in XML, and as these are just text files anyway, XML would be an upgrade for readability and would minimize human error
- Literally all we need is a type of language that minimizes the chance for human error to come into play, and makes tools easier to make. 

I would like to leave all files that *don't have tools being developed for them* in their current .txt state, but the hope is that with later releases, tools like the ESMB and endless sky map editor will be packaged in releases of the game to make content creating and modding easier to a large amount of players, meaning normal players should never have to crack open the xml (or whatever language) files.

For this, I honestly don't care if the size of the game is increased greatly. (Reason: Any PC capable of running OGL 3.0+ is already capable of running several instances of endless sky without any issue, and even on shitty internet that's in the kbps ranges, Endless Sky downloads in a few minutes. We have a lot of wiggle room and play without completely killing efficiency.) I just want content creating to be as simple and streamlined as possible for modders and regular players (also for myself).

**Comment by kikotheexile** [comment]
TL;DR let's make a decision and implement it. I don't think it really matters too much what language we decide on, especially as there are conversion tools out there so we can always change our collective mind.

**Comment by shitwolfymakes** [comment]
> I would like to leave all files that don't have tools being developed for them in their current .txt state

The problem with this is that the code currently parses all data files in one way, so changing the format for one set of files means changing it for all, having to deal with 2 systems that read data from files.

All or nothing, XML has everything we need

**Comment by kikotheexile** [comment]
There are so many arguments that I could make to the contrary, but it's all shit you already know. (Games like SWEAW read and parse from several file types (XML and Lua), basically everything can read .txt files so parsing from that isn't a huge deal. So on, so on, so on. )

But the long and short of it is: anything to make creation of content easier for everyone - from illiterate newbie to skilled veteran - is the end goal here. 

If we *must* go all in, then perhaps we should set up a *plan for going all in.* Update the ES Editor to read from XML (and not be broken), work with Darcy on fixing the ship builder and packaging that as an application as well (it being web based is fine, too, but you get my point)

The main concerns here are two-fold. Programmers want the file type to be easy for them to read and QC, and I want idiots like myself to be able to just plug in and fix things on the fly.

**Comment by kikotheexile** [comment]
On Discord we were able to sway @lukearndt that XML is the best option we have for what we need, even though we all agree that we wish we could use YAML or another language. So I think we will be moving forward with this. 

I know no one really likes the decision, and the feeling is shared among everyone pro and anti XML alike, it is literally just the best option for the tools that are to be developed.

So the issue will once again return to the bounty - we just need someone to make it so and convert the files over.

**Comment by kikotheexile** [comment]
@MCOfficer is also now begrudgingly on board with this as well, citing:

>i'm convinced... not by the arguments against yaml/json, but rather by the lack of arguments against xml. whatever

**Comment by shitwolfymakes** [comment]
Now that this is decided, I'll start going through everything, and draw up examples of converted files for people to work with. I'll attach them as I complete them

**Comment by shitwolfymakes** [comment]
Due to the overwhelming lack of support for this on Endless Sky, I'm closing this pull request for both ES and ESCW, so that I can continue to support both games at the same time

**Comment by MCOfficer** [comment]
To bridge this gap, i'm working on a python module that is supposed to act as robust and (eventually) high-level parser and writer. The goal is to have a library that doesn't differ a lot from an xml-parser as far as the program using it is concerned.
https://github.com/EndlessSkyCommunity/EndlessParser

**Comment by kikotheexile** [comment]
eeeeeeeeeeeeeeeeeeeeeeeerrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr
ok
